TLM/TMPE.API/Implementations.cs

```diff
@@ -4,13 +4,29 @@ namespace TrafficManager.API {
     using TrafficManager.API.Notifier;
     using System.Linq;
     using TrafficManager.API.UI;
+    using ColossalFramework.Plugins;
+    using System.Reflection;
 
     public static class Implementations {
         private static Type constantsType_;
         private static IManagerFactory managerFactory_;
         private static INotifier notifier_;
         private static IUIFactory uiFactory_;
 
+        static Implementations() {
+            // reset on hot reload or game reload.
+            PluginManager.instance.eventPluginsStateChanged += Reset;
+            PluginManager.instance.eventPluginsChanged += Reset;
+            LoadingManager.instance.m_levelPreLoaded += Reset;
+        }
+
+        private static void Reset() {
+            constantsType_ = null;
+            managerFactory_ = null;
+            notifier_ = null;
+            uiFactory_ = null;
```

Comment:
Hmm, when it will unsubscribe from event? If we hotreload the mod, API also gets reloaded 🤔

Reply:
if you hot reload only `TrafficManager.dll` but not `TMPE.API.dll` then it will work.
why would you unsubscribe from the events?

Reply:
alternatively I can do this if you don't like events:
```diff
--- a/TLM/TLM/Lifecycle/TMPELifecycle.cs
+++ b/TLM/TLM/Lifecycle/TMPELifecycle.cs
@@ -222,6 +222,7 @@ namespace TrafficManager.Lifecycle {

                 InGameHotReload = InGameOrEditor();
                 if (InGameHotReload) {
+                    TrafficManager.API.Implementations.Reset();
                     Preload();
                     SerializableDataExtension.Load();
                     Load();
```

Reply:
lets go with the other approach then. easier to manage lifecycle.

Reply:
ok

Reply:
Done